### PR TITLE
split escrow between normal tokens and tokens undergoing debonding

### PR DESCRIPTION
**Pull request by pro-wh**
We'll need this for rewards, so that we can increase the value of the escrow shares that aren't undergoing debonding.

blocks #2288

cc #2290, which also changes slashing to take a fraction. That PR uses a single shared denominator, which is fine.

### CODE REVIEW

File: go/staking/api/api.go [hunk 1]
```diff
@@ -332,48 +332,69 @@ func MoveUpTo(dst, src, n *Quantity) (*Quantity, error) {
...
 
-	return issuedShares, nil
+	return q, nil
 }
 
-// TokensForShares computes the amount of tokens to be received for shares in the
-// given escrow account.
-func TokensForShares(acc *EscrowAccount, amount *Quantity) (*Quantity, error) {
-	if amount.IsZero() || acc.Balance.IsZero() || acc.TotalShares.IsZero() {
+func (p *SharePool) Deposit(shareDst, tokenSrc, tokenAmount *Quantity) error {
```
Comment (kostko) [inline go/staking/api/api.go]: Documentation comment? Mention that in case of errors the pool can still be modified.

File: go/staking/api/api.go [hunk 2]
```diff
@@ -384,19 +405,40 @@ func TokensForShares(acc *EscrowAccount, amount *Quantity) (*Quantity, error) {
...
 	//       lost due to loss of precision.
-	if err := q.Quo(&acc.TotalShares); err != nil {
+	if err := q.Quo(&p.TotalShares); err != nil {
 		return nil, err
 	}
 
 	return q, nil
 }
 
+func (p *SharePool) Withdraw(tokenDst, shareSrc, shareAmount *Quantity) error {
```
Comment (kostko) [inline go/staking/api/api.go]: Documentation comment?

File: go/tendermint/apps/staking/staking.go [hunk 3]
```diff
@@ -112,48 +112,26 @@ func (app *stakingApplication) onEpochChange(ctx *abci.Context, epoch epochtime.
 			escrow = state.Account(e.EscrowID)
 		}
 
-		// Compute amount of debonded tokens.
-		tokens, err := staking.TokensForShares(&escrow.Escrow, &deb.Shares)
-		if err != nil {
-			app.logger.Error("failed to compute amount of tokens from shares",
+		var tokens staking.Quantity
+		if err := escrow.Escrow.Debonding.Withdraw(&tokens, &deb.Shares, &deb.Shares); err != nil {
```
Comment (kostko) [inline go/tendermint/apps/staking/staking.go]: Is it ok if `Withdraw` is called with `srcShares` and `shareAmount` being the same pointer? Based on that function code it seems like there could be some unintended corruption.

Comment (pro-wh) [reply go/tendermint/apps/staking/staking.go]: ah shoot no

File: go/staking/api/api.go (hunk 1 above)
Comment (pro-wh) [reply go/staking/api/api.go]: ok

File: go/staking/api/api.go (hunk 2 above)
Comment (pro-wh) [reply go/staking/api/api.go]: ok

File: go/tendermint/apps/staking/staking.go (hunk 3 above)
Comment (pro-wh) [reply go/tendermint/apps/staking/staking.go]: added Clone ):

Comment (pro-wh) [reply go/tendermint/apps/staking/staking.go]: oops, had some other instances. fixed those too

Comment (pro-wh) [reply go/tendermint/apps/staking/staking.go]: added tests in #2288

**Pull request merged**